streamrouter/consul.py

```diff
@@ -1,485 +1,552 @@
-        return self.__get_services(self.__arrow_asns_services, tag)
+        try:
+            while True:
+                service = itertor_next_func(iterator)
+                if service is None:
+                    break
+
+                services.append(service_factory(service))
+        finally:
+            iterator_free_func(iterator)
+
+        return services
+
+    def get_rtspcon_services(self):
+        """
+        Get a list of all RTSP Connector services.
+        """
+        return self.get_services(
+            lib.srl__consul__get_all_rtspcon_services,
+            lib.srl__rtspcon_services__next,
+            lib.srl__rtspcon_services__free,
+            RtspconService)
+
+    def get_hls_edge_services(self):
+        """
+        Get a list of all HLS edge services.
+        """
+        return self.get_services(
+            lib.srl__consul__get_all_hls_edge_services,
+            lib.srl__hls_edge_services__next,
+            lib.srl__hls_edge_services__free,
+            HlsEdgeService)
+
+    def get_mp4_edge_services(self):
+        """
+        Get a list of all MP4 edge services.
+        """
+        return self.get_services(
+            lib.srl__consul__get_all_mp4_edge_services,
+            lib.srl__mp4_edge_services__next,
+            lib.srl__mp4_edge_services__free,
+            Mp4EdgeService)
+
+    def get_mjpeg_proxy_services(self):
+        """
+        Get a list of all MJPEG proxy services.
+        """
+        return self.get_services(
+            lib.srl__consul__get_all_mjpeg_proxy_services,
+            lib.srl__mjpeg_proxy_services__next,
+            lib.srl__mjpeg_proxy_services__free,
+            MjpegProxyService)
+
+    def get_arrow_asns_services(self):
+        """
+        Get a list of all Arrow ASNS services.
+        """
+        return self.get_services(
+            lib.srl__consul__get_all_arrow_asns_services,
+            lib.srl__arrow_asns_services__next,
+            lib.srl__arrow_asns_services__free,
+            ArrowAsnsService)
 
     def add_update_callback(self, cb):
-        self.__update_callbacks.add(cb)
+        """
+        Add a given update callback. The callback will be invoked whenever
+        the consul service gets updated.
+        """
+        def callback(err):
+            cb()
+
+        ncb = lib.UPDATE_CALLBACK(callback)
+
+        assert self.raw_ptr is not None
 
-    def remove_update_callback(self, cb):
-        self.__update_callbacks.remove(cb)
+        token = lib.srl__consul__add_update_callback(self.raw_ptr, ncb)
+
+        self.update_callbacks[token] = ncb
+
+        return token
+
+    def remove_update_callback(self, token):
+        """
+        Remove an update callback corresponding to a given token.
+        """
+        assert self.raw_ptr is not None
+
+        lib.srl__consul__remove_update_callback(self.raw_ptr, token)
+
+        try:
+            del self.update_callbacks[token]
```

Comment:
what about to use `self.update_callbacks.pop(token, None)` without try/except block